[dtensor][debug] force visualize_sharding not to print for empty tensors

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121392
* #120265
* __->__ #121217
* #121382
* #121385

**Summary**
Current `visualize_sharding` code cannot print for empty DTensor objects which leads to an exception. This PR skips the print logic if the DTensor passed in has 0 element.
<img width="2165" alt="Pasted Graphic" src="https://github.com/pytorch/pytorch/assets/12968408/fa40b5e7-dad7-4d3a-a485-6a18067320ff">


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

XilunWu committed:
[dtensor][debug] force visualize_sharding not to print for empty tensors

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/debug/visualize_sharding.py:
```diff
@@ -135,6 +135,9 @@ def visualize_sharding(dtensor, prompt=""):
     Visualizes sharding in 1D-2D dtensors
     Requires tabulate, install with `pip install tabulate`
     """
+    if (dtensor.numel() == 0):  # we do not print for empty tensors
```
hmmm is this right? this only print empty DTensor but not the case where it's a DTensor that some ranks have data but some ranks have empty tensor?

XilunWu replied:
yeah, you're right. 2 things to address:
1) if the DTensor is empty, visualize_sharding will throw exception when calculating the offset. So I skip if this is the case in code in this PR.
2) if the DTensor is placed on a submesh, visualize_sharding will throw exception when calling `get_local_rank()` because non-participating ranks have empty `_dim_group_infos`. So I skip on all non-participating ranks. This change is in #121382

XilunWu committed:
Update on "[dtensor][debug] force visualize_sharding not to print for empty tensors"


**Summary**
Current `visualize_sharding` code cannot print for empty DTensor objects which leads to an exception. This PR skips the print logic if the DTensor passed in has 0 element.
<img width="2165" alt="Pasted Graphic" src="https://github.com/pytorch/pytorch/assets/12968408/fa40b5e7-dad7-4d3a-a485-6a18067320ff">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][debug] force visualize_sharding not to print for empty tensors"


**Summary**
Current `visualize_sharding` code cannot print for empty DTensor objects which leads to an exception. This PR skips the print logic if the DTensor passed in has 0 element.
<img width="2165" alt="Pasted Graphic" src="https://github.com/pytorch/pytorch/assets/12968408/fa40b5e7-dad7-4d3a-a485-6a18067320ff">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][debug] force visualize_sharding not to print for empty tensors"


**Summary**
Current `visualize_sharding` code cannot print for empty DTensor objects which leads to an exception. This PR skips the print logic if the DTensor passed in has 0 element.
<img width="2165" alt="Pasted Graphic" src="https://github.com/pytorch/pytorch/assets/12968408/fa40b5e7-dad7-4d3a-a485-6a18067320ff">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][debug] force visualize_sharding not to print for empty tensors"


**Summary**
Current `visualize_sharding` code cannot print for empty DTensor objects which leads to an exception. This PR skips the print logic if the DTensor passed in has 0 element.
<img width="2165" alt="Pasted Graphic" src="https://github.com/pytorch/pytorch/assets/12968408/fa40b5e7-dad7-4d3a-a485-6a18067320ff">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][debug] force visualize_sharding not to print for empty tensors"


**Summary**
Current `visualize_sharding` code cannot print for empty DTensor objects which leads to an exception. This PR skips the print logic if the DTensor passed in has 0 element.
<img width="2165" alt="Pasted Graphic" src="https://github.com/pytorch/pytorch/assets/12968408/fa40b5e7-dad7-4d3a-a485-6a18067320ff">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/examples/visualize_sharding_example.py:
```diff
@@ -82,3 +82,12 @@
 """
+
+# Example 6: empty shard
+tensor = torch.empty(0)
```
hmmm I think this should not be an example, it's more like edge case handling and user should not be aware of this at all.

wanchaol reviewed (approved):
I think we should remove the examples as inlined comment, otherwise lgtm

XilunWu committed:
Update on "[dtensor][debug] force visualize_sharding not to print for empty tensors"


**Summary**
Current `visualize_sharding` code cannot print for empty DTensor objects which leads to an exception. This PR skips the print logic if the DTensor passed in has 0 element.
<img width="2165" alt="Pasted Graphic" src="https://github.com/pytorch/pytorch/assets/12968408/fa40b5e7-dad7-4d3a-a485-6a18067320ff">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge

XilunWu commented:
@pytorchbot merge
